### ISSUE TITLE
LSTM Batch Normalization with Multi GPU

### ISSUE THREAD

**Issue by yudeshui** [apache/incubator-mxnet#4664]
I have implemented LSTM with Batch Normalization as followed:

      def lstm(self, prefix, num_hidden, indata, prev_state, param, seqidx, layeridx, dropout=0.):
        if dropout > 0.:
            indata = mx.sym.Dropout(data=indata, p=dropout)
        i2h = mx.sym.FullyConnected(data=indata,
                                    weight=param.i2h_weight,
                                    num_hidden=num_hidden * 4,
                                    no_bias=True,
                                    name=prefix + "t%d_l%d_i2h" % (seqidx, layeridx))
        i2h = self.BatchNorm(
            name=prefix + "t%d_l%d_i2h_bn" % (seqidx, layeridx),
            data=i2h,
            gamma=param.i2h_gamma,
            num_channel=num_hidden * 4
        )
        h2h = mx.sym.FullyConnected(data=prev_state.h,
                                    weight=param.h2h_weight,
                                    num_hidden=num_hidden * 4,
                                    no_bias=True,
                                    name=prefix + "t%d_l%d_h2h" % (seqidx, layeridx))
        h2h = self.BatchNorm(
            name=prefix + "t%d_l%d_h2h_bn" % (seqidx, layeridx),
            data=h2h,
            gamma=param.h2h_gamma,
            beta=param.beta,
            num_channel=num_hidden * 4
        )
        gates = i2h + h2h
        slice_gates = mx.sym.SliceChannel(gates, num_outputs=4,
                                          name=prefix + "t%d_l%d_slice" % (seqidx, layeridx))
        in_gate = mx.sym.Activation(slice_gates[0], act_type="sigmoid")
        in_transform = mx.sym.Activation(slice_gates[1], act_type="tanh")
        forget_gate = mx.sym.Activation(slice_gates[2], act_type="sigmoid")
        out_gate = mx.sym.Activation(slice_gates[3], act_type="sigmoid")
        next_c = (forget_gate * prev_state.c) + (in_gate * in_transform)
        next_h = out_gate * mx.sym.Activation(
            self.BatchNorm(
                name=prefix + 't%d_l%d_c_bn' % (seqidx, layeridx),
                data=next_c,
                gamma=param.c_gamma,
                beta=param.c_beta,
                num_channel=num_hidden
            ),
            act_type="tanh"
        )
        return LSTMState(c=next_c, h=next_h)

    def BatchNorm(self, name, data, gamma, beta=None, **kargs):
        net = data

        if not self.bn:
            return net

        if self.minibatch:
            num_channel = kargs.pop('num_channel')
            net = mx.symbol.Reshape(net, shape=(-1, self.num_subject * num_channel))
            net = mx.symbol.BatchNorm(
                name=name + '_norm',
                data=net,
                fix_gamma=True,
                momentum=0.9,
                attr={'wd_mult': '0', 'lr_mult': '0'}
            )
            net = mx.symbol.Reshape(data=net, shape=(-1, num_channel))
        else:
            net = mx.symbol.BatchNorm(
                name=name + '_norm',
                data=net,
                fix_gamma=True,
                momentum=0.9,
                attr={'wd_mult': '0', 'lr_mult': '0'}
            )
        net = mx.symbol.broadcast_mul(net, gamma)
        if beta is not None:
            net = mx.symbol.broadcast_plus(net, beta)
        return net

Setting the gpu as (0), the train accuracy and validation accuracy are
```
[2017-01-13 13:19:22.762397] INFO: root: Epoch[0] Batch [50]	Speed: 19871.61 samples/sec	Train-accuracy[g]=0.084730
[2017-01-13 13:19:24.705928] INFO: root: Epoch[0] Batch [100]	Speed: 20584.81 samples/sec	Train-accuracy[g]=0.240150
[2017-01-13 13:19:26.896294] INFO: root: Epoch[0] Train-accuracy[g]=0.529115
[2017-01-13 13:19:26.896477] INFO: root: Epoch[0] Time cost=6.213
[2017-01-13 13:19:26.953329] INFO: root: Saved checkpoint to ".cache/srep-ninaprodb1-lowpass-one-fold-intra-subject-0-v2017.1.8-lstm-last1-rw4-h64-p0.2-bn-nopretrain/model-0001.params"
[2017-01-13 13:19:30.201415] INFO: root: Epoch[0] Validation-accuracy[g]=0.616754
[2017-01-13 13:19:32.927509] INFO: root: Epoch[1] Batch [50]	Speed: 15228.44 samples/sec	Train-accuracy[g]=0.696912
[2017-01-13 13:19:34.690835] INFO: root: Epoch[1] Batch [100]	Speed: 22690.98 samples/sec	Train-accuracy[g]=0.773525
[2017-01-13 13:19:36.454076] INFO: root: Epoch[1] Train-accuracy[g]=0.814401
[2017-01-13 13:19:36.454352] INFO: root: Epoch[1] Time cost=6.236
[2017-01-13 13:19:39.748917] INFO: root: Epoch[1] Validation-accuracy[g]=0.783105
[2017-01-13 13:19:42.132682] INFO: root: Epoch[2] Batch [50]	Speed: 17607.58 samples/sec	Train-accuracy[g]=0.850098
[2017-01-13 13:19:44.452310] INFO: root: Epoch[2] Batch [100]	Speed: 17247.31 samples/sec	Train-accuracy[g]=0.866325
[2017-01-13 13:19:46.757555] INFO: root: Epoch[2] Train-accuracy[g]=0.887682
[2017-01-13 13:19:46.757761] INFO: root: Epoch[2] Time cost=6.992
[2017-01-13 13:19:50.001279] INFO: root: Epoch[2] Validation-accuracy[g]=0.771694
[2017-01-13 13:19:52.206201] INFO: root: Epoch[3] Batch [50]	Speed: 19019.27 samples/sec	Train-accuracy[g]=0.905343
[2017-01-13 13:19:53.960932] INFO: root: Epoch[3] Batch [100]	Speed: 22800.01 samples/sec	Train-accuracy[g]=0.916825
[2017-01-13 13:19:55.716543] INFO: root: Epoch[3] Train-accuracy[g]=0.924427
[2017-01-13 13:19:55.716728] INFO: root: Epoch[3] Time cost=5.699
[2017-01-13 13:19:59.454541] INFO: root: Epoch[3] Validation-accuracy[g]=0.805141
[2017-01-13 13:20:01.945854] INFO: root: Epoch[4] Batch [50]	Speed: 16898.57 samples/sec	Train-accuracy[g]=0.934020
[2017-01-13 13:20:04.016064] INFO: root: Epoch[4] Batch [100]	Speed: 19325.80 samples/sec	Train-accuracy[g]=0.938825
[2017-01-13 13:20:05.731813] INFO: root: Epoch[4] Train-accuracy[g]=0.946172
[2017-01-13 13:20:05.732054] INFO: root: Epoch[4] Time cost=6.261
[2017-01-13 13:20:09.077874] INFO: root: Epoch[4] Validation-accuracy[g]=0.802944
```
Setting the gpu as (0,1), the train accuracy and valiadation accuracy are
```
[2017-01-13 13:36:09.765492] INFO: root: Epoch[0] Batch [50]	Speed: 14728.07 samples/sec	Train-accuracy[g]=0.081176
[2017-01-13 13:36:12.639381] INFO: root: Epoch[0] Batch [100]	Speed: 13920.82 samples/sec	Train-accuracy[g]=0.240725
[2017-01-13 13:36:15.171587] INFO: root: Epoch[0] Train-accuracy[g]=0.530859
[2017-01-13 13:36:15.171819] INFO: root: Epoch[0] Time cost=8.203
[2017-01-13 13:36:15.224567] INFO: root: Saved checkpoint to ".cache/srep-ninaprodb1-lowpass-one-fold-intra-subject-0-v2017.1.8-lstm-last1-rw4-h64-p0.2-bn-nopretrain-2gpu/model-0001.params"
[2017-01-13 13:36:18.943427] INFO: root: Epoch[0] Validation-accuracy[g]=0.020363
[2017-01-13 13:36:21.583001] INFO: root: Epoch[1] Batch [50]	Speed: 15811.54 samples/sec	Train-accuracy[g]=0.690662
[2017-01-13 13:36:24.217206] INFO: root: Epoch[1] Batch [100]	Speed: 15187.10 samples/sec	Train-accuracy[g]=0.771225
[2017-01-13 13:36:26.820433] INFO: root: Epoch[1] Train-accuracy[g]=0.817656
[2017-01-13 13:36:26.820705] INFO: root: Epoch[1] Time cost=7.861
[2017-01-13 13:36:30.832853] INFO: root: Epoch[1] Validation-accuracy[g]=0.020363
[2017-01-13 13:36:33.546381] INFO: root: Epoch[2] Batch [50]	Speed: 15131.85 samples/sec	Train-accuracy[g]=0.846225
[2017-01-13 13:36:36.063439] INFO: root: Epoch[2] Batch [100]	Speed: 15895.43 samples/sec	Train-accuracy[g]=0.868600
[2017-01-13 13:36:38.660459] INFO: root: Epoch[2] Train-accuracy[g]=0.888021
[2017-01-13 13:36:38.660775] INFO: root: Epoch[2] Time cost=7.812
[2017-01-13 13:36:42.269283] INFO: root: Epoch[2] Validation-accuracy[g]=0.020363
[2017-01-13 13:36:44.975286] INFO: root: Epoch[3] Batch [50]	Speed: 15265.43 samples/sec	Train-accuracy[g]=0.902083
[2017-01-13 13:36:47.547094] INFO: root: Epoch[3] Batch [100]	Speed: 15555.77 samples/sec	Train-accuracy[g]=0.914100
[2017-01-13 13:36:49.974006] INFO: root: Epoch[3] Train-accuracy[g]=0.925078
[2017-01-13 13:36:49.974301] INFO: root: Epoch[3] Time cost=7.678
[2017-01-13 13:36:53.442460] INFO: root: Epoch[3] Validation-accuracy[g]=0.020363
[2017-01-13 13:36:55.981809] INFO: root: Epoch[4] Batch [50]	Speed: 16373.85 samples/sec	Train-accuracy[g]=0.933480
[2017-01-13 13:36:58.740740] INFO: root: Epoch[4] Batch [100]	Speed: 14500.27 samples/sec	Train-accuracy[g]=0.939650
[2017-01-13 13:37:01.364774] INFO: root: Epoch[4] Train-accuracy[g]=0.943203
[2017-01-13 13:37:01.365013] INFO: root: Epoch[4] Time cost=7.906
[2017-01-13 13:37:04.896343] INFO: root: Epoch[4] Validation-accuracy[g]=0.020363
```
Any idea why ?
Thanks :)

**Comment by hengck23** [comment]
@yudeshui 
Hi, I am trying to make caffe scale layer from mxnet operation.
In particular, i can use:
```
net = mx.symbol.broadcast_mul(net, gamma)
net = mx.symbol.broadcast_plus(net, beta)
```
However,  if i use net.list_arguments(),  gamma and beta are not in the list.
(meaning that gamma,beta are not learn-able parameters)

Do you know how to solve it?

**Comment by phunterlau** [comment]
This issue is closed due to lack of activity in the last 90 days. Feel free to reopen if this is still an active issue. Thanks!
